components/concept/fetch/src/main/java/mozilla/components/concept/fetch/Client.kt

```diff
@@ -62,3 +66,31 @@ abstract class Client {
         "Connection" to "keep-alive"
     )
 }
+
+/**
+ * "Fetches" a base64 encoded data URI.
+ *
+ * @param request The [Request] for the data URI.
+ * @return The generated [Response] including the decoded bytes as body.
+ */
+@Suppress("TooGenericExceptionCaught")
+fun fetchDataUri(request: Request): Response {
+    return try {
+        val dataUri = request.url
+        val contentType = dataUri.substringAfter("data:").substringBefore(";base64")
+
+        val bytes = Base64.decode(dataUri.substring(dataUri.lastIndexOf(',') + 1), Base64.DEFAULT)
+        val headers = MutableHeaders(
+            CONTENT_TYPE_HEADER to contentType,
+            CONTENT_LENGTH_HEADER to bytes.size.toString()
+        )
```

Comment:
Looking at [RFC 2397](https://tools.ietf.org/html/rfc2397) and [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs) I wonder if we should be a bit more protective regarding optional elements:
* If the data uri doesn't specify base64 encoding then bail out early?
* If the content type is not specified then do not add header for it?

Reply:
OK, added another test case and made this more robust too.